File: app/src/commonMain/kotlin/com/crosspaste/app/CrossPasteWebService.kt
package com.crosspaste.app

import com.crosspaste.net.ResourcesClient
import io.github.oshai.kotlinlogging.KotlinLogging
import io.ktor.util.collections.ConcurrentMap
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

@Serializable
data class WebLocale(
    val code: String,
    val label: String,
    val path: String,
)

@Serializable
data class WebMeta(
    val locales: List<WebLocale>,
)

class CrossPasteWebService(
    private val appUrls: AppUrls,
    private val resourcesClient: ResourcesClient,
) {

    private val logger = KotlinLogging.logger {}

    private val json =
        Json {
            ignoreUnknownKeys = true
        }

    private var localePathMap: Map<String, String> = ConcurrentMap()

    suspend fun refresh() {
        runCatching {
            val metaUrl = "${appUrls.homeUrl}/api/meta.json"
            resourcesClient
                .request(metaUrl)
                .getOrNull()
                ?.let { response ->
                    val text = response.getBodyAsText()
                    val meta = json.decodeFromString<WebMeta>(text)
                    localePathMap = meta.locales.associate { it.code to it.path }
                    logger.info { "Web locale config loaded: ${localePathMap.keys}" }
                }
        }.onFailure { e ->
            logger.warn { "Failed to fetch web locale config: ${e.message}" }
        }
    }

    fun getWebUrl(
        language: String,
        path: String = "",
    ): String {
        val localePath = resolveLocalePath(language)
        return "${appUrls.homeUrl}$localePath$path"
    }

    private fun resolveLocalePath(language: String): String {
        val map = localePathMap
        if (map.isNotEmpty()) {
            return map[language] ?: map["en"] ?: "/en/"
        }
        // Fallback before meta.json loads
        return if (language == "zh") "/" else "/en/"
    }
}

File: web/src/shared/app/cross-paste-web-service.ts
import { AppUrls } from "./app-urls";
import { getActiveLanguage } from "@/shared/i18n/i18n-core";

/**
 * Mirrors the Kotlin [CrossPasteWebService] — resolves the locale path
 * segment of https://crosspaste.com so we can deep-link users to the
 * localized landing / download page.
 *
 * The desktop app fetches `/api/meta.json` and caches `localePathMap`;
 * we do the same here (lazy, in-memory) with a hard-coded fallback for
 * the cold-start case (no network or before refresh completes).
 */

interface WebLocale {
  code: string;
  label: string;
  path: string;
}

interface WebMeta {
  locales: WebLocale[];
}

let localePathMap: Record<string, string> = {};

async function refresh(): Promise<void> {
  try {
    const response = await fetch(`${AppUrls.homeUrl}/api/meta.json`, { cache: "no-cache" });
    if (!response.ok) return;
    const meta = (await response.json()) as WebMeta;
    localePathMap = Object.fromEntries(meta.locales.map((l) => [l.code, l.path]));
  } catch (e) {
    console.warn("[CrossPasteWebService] Failed to fetch web meta:", e);
  }
}

function resolveLocalePath(language: string): string {
  const map = localePathMap;
  if (Object.keys(map).length > 0) {
    return map[language] ?? map["en"] ?? "/en/";
  }
  // Fallback before meta.json loads, matching the desktop default.
  return language === "zh" ? "/" : "/en/";
}

export const CrossPasteWebService = {
  refresh,

  /** Build a localized URL like `https://crosspaste.com/en/download`. */
  getWebUrl(language: string, path = ""): string {
    return `${AppUrls.homeUrl}${resolveLocalePath(language)}${path}`;
  },

  /** Shortcut that reads the active UI language from storage. */
  async getLocalizedUrl(path = ""): Promise<string> {
    return this.getWebUrl(await getActiveLanguage(), path);
  },
};
